substrate/client/network/src/litep2p/shim/notification/tests/fuzz.rs

```diff
@@ -0,0 +1,358 @@
+		.map(|_| {
+			let id = PeerId::random();
+			known_nodes.insert(id, State::Disconnected);
+			id
+		})
+		.collect();
+
+	let peerstore = Peerstore::new(bootnodes);
+	let peer_store_handle = peerstore.handle();
+
+	let (mut peerset, to_peerset) = Peerset::new(
+		ProtocolName::from("/notif/1"),
+		Uniform::new_inclusive(0, 25).sample(&mut rng),
+		Uniform::new_inclusive(0, 25).sample(&mut rng),
+		Uniform::new_inclusive(0, 10).sample(&mut rng) == 0,
+		(0..Uniform::new_inclusive(0, 2).sample(&mut rng))
+			.map(|_| {
+				let id = PeerId::random();
+				known_nodes.insert(id, State::Disconnected);
+				reserved_nodes.insert(id);
+				id
+			})
+			.collect(),
+		Default::default(),
+		Arc::new(peerstore_handle_test()),
+	);
+
+	tokio::spawn(peerstore.run());
+
+	// list of nodes the user of `peerset` assumes it's connected to
+	//
+	// always a subset of `known_nodes`.
+	let mut connected_nodes = HashSet::<PeerId>::new();
+
+	// list of nodes the user of `peerset` called `incoming` with and that haven't been
+	// accepted or rejected yet.
+	let mut incoming_nodes = HashMap::<usize, PeerId>::new();
+
+	// next id for incoming connections.
+	let mut next_incoming_id = 0usize;
+
+	// peers for whom substreams are opening
+	let mut opening: HashSet<PeerId> = HashSet::new();
+
+	// peers for whom substream is closing
+	let mut closing: HashSet<PeerId> = HashSet::new();
+
+	// peers who are connected
+	let mut connected: HashSet<PeerId> = HashSet::new();
+
+	// peers who are backed off
+	let mut backed_off: HashMap<PeerId, Instant> = HashMap::new();
+
+	// The loop below is effectively synchronous, so for `PeerStore` & `ProtocolController`
+	// runners, spawned above, to advance, we use `spawn_blocking`.
+	let _ = tokio::task::spawn_blocking(move || {
+		// PRNG to use in `spawn_blocking` context.
+		let mut rng = rand::thread_rng();
+
+		// perform a certain number of actions while checking that the state is consistent.
+		//
+		// if we reach the end of the loop, the run has succeeded
+		for _ in 0..5000 {
+			// peer we are working with
+			let mut current_peer = None;
+
+			// current event for state transition validation
+			let mut current_event = None;
+
+			// last peer state for allowed event validation
+			let mut last_state = None;
+
+			// each of these weights corresponds to an action that we may perform
+			let action_weights = [150, 90, 90, 30, 30, 1, 1, 4, 4, 90];
+
+			match WeightedIndex::new(&action_weights).unwrap().sample(&mut rng) {
+				0 => match peerset.next().now_or_never() {
+					Some(Some(PeersetNotificationCommand::OpenSubstream { peers })) => {
+						for peer in peers {
+							assert!(opening.insert(peer));
+						}
+					},
+					Some(Some(PeersetNotificationCommand::CloseSubstream { peers })) => {
+						for peer in peers {
+							assert!(closing.insert(peer));
+						}
+					},
+					Some(None) => panic!("peerset exited"),
+					None => {},
+				},
```

Comment:
`PsmConnect` & `PsmDrop` events are not used.

Reply:
AFAIR they didn't make sense for this `Peerset` but I'll have to recheck

Reply:
Resolving, see the comment above